//specify the proto version
syntax = "proto3";
//the name(or empty)
package groupChat;

service ChatServer {
    rpc chatFunction(ChatInput) returns (ChatOutput) {}
    rpc getMessages(ChatInput) returns (stream ChatMessage) {}
    rpc syncMessage(ChatServerSyncRequest) returns (ChatServerSyncResponse);
    rpc sendMessage(ChatServerRequest) returns (ChatServerResponse);
    rpc probe(Empty) returns (Empty);
}

message ChatInput {
    //type 1: login, 2: join, 3: chat, 4: like, 5: dislike, 6: history
    string userName = 1;
    string groupName = 2;
    int32 type = 3;
    string message = 4;
    int32 messageId = 5;
    string uuid = 6;
    int32 serverId = 7;
}

message ChatOutput {
    string status = 1;
    repeated ChatMessage messages = 2;
    repeated string user = 3;
}

message ChatMessage{
    int32 id = 1;
    string user = 2;
    string content = 3;
    int32 numberOfLikes = 4;
}

message Empty{

}

message ChatServerRequest {
    repeated int32 vector = 1;
    ChatInput request = 2;
    int32 server_id = 3;
    int32 mode = 4;
}

message ChatServerResponse {
    string status = 1;
}

message ChatServerSyncRequest {
    repeated int32 vector = 1;
    int32 server_id = 2;
}

message ChatServerSyncResponse {
    string status = 1;
}